research/deep_speech/decoder.py

```diff
@@ -0,0 +1,206 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Deep speech decoder."""
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+from nltk.metrics import distance
+from six.moves import xrange
+import tensorflow as tf
+
+
+class DeepSpeechDecoder(object):
+  """Basic decoder class from which all other decoders inherit.
+
+  Implements several helper functions. Subclasses should implement the decode()
+  method.
+  """
+
+  def __init__(self, labels, blank_index=28, space_index=27):
+    """Decoder initialization.
+
+    Arguments:
+      labels (string): mapping from integers to characters.
+      blank_index (int, optional): index for the blank '_' character.
+        Defaults to 0.
+      space_index (int, optional): index for the space ' ' character.
+        Defaults to 28.
+    """
+    # e.g. labels = "[a-z]' _"
+    self.labels = labels
+    self.int_to_char = dict([(i, c) for (i, c) in enumerate(labels)])
+    self.blank_index = blank_index
+    self.space_index = space_index
+
+  def convert_to_strings(self, sequences, sizes=None):
+    """Given a list of numeric sequences, returns the corresponding strings."""
+    strings = []
+    for x in xrange(len(sequences)):
+      seq_len = sizes[x] if sizes is not None else len(sequences[x])
+      string = self._convert_to_string(sequences[x], seq_len)
+      strings.append(string)
+    return strings
+
+  def _convert_to_string(self, sequence, sizes):
+    return ''.join([self.int_to_char[sequence[i]] for i in range(sizes)])
+
+  def process_strings(self, sequences, remove_repetitions=False):
+    """Process strings.
+
+    Given a list of strings, removes blanks and replace space character with
+    space. Option to remove repetitions (e.g. 'abbca' -> 'abca').
+
+    Arguments:
+      sequences: list of 1-d array of integers
+      remove_repetitions (boolean, optional): If true, repeating characters
+        are removed. Defaults to False.
+
+    Returns:
+      The processed string.
+    """
+    processed_strings = []
+    for sequence in sequences:
+      string = self.process_string(remove_repetitions, sequence).strip()
+      processed_strings.append(string)
+    return processed_strings
+
+  def process_string(self, remove_repetitions, sequence):
+    """Process each given sequence."""
+    seq_string = ''
```

Comment:
In case your input string is very long, you should use list to hold individual char, and join them at the end with "". Python string is immutable object. concat string, means allocate new string and delete the old one.

https://waymoot.org/home/python_string/